src/_posts/2017-08-04-splittable-do-fn.md

```diff
@@ -0,0 +1,449 @@
+see [Non-IO examples](http://s.apache.org/splittable-do-fn#heading=h.5cep9s8k4fxv))*.
+
+### Execution of a restriction and data consistency
+
+One of the most important parts of the Splittable `DoFn` design is related to
+how it achieves data consistency while splitting. For example, while the runner
+is preparing to split the restriction of an active `@ProcessElement` call, how
+can it be sure that the call has not concurrently progressed past the point of
+splitting?
+
+This is achieved by requiring the processing of a restriction to follow a
+certain pattern. We think of a restriction as a sequence of *blocks* -
+elementary indivisible units of work, identified by a *position*. A
+`@ProcessElement` call processes the blocks one by one, first *claiming* the
+block's position to atomically check if it's still within the range of the
+restriction, until the whole restriction is processed.
+
+The diagram below illustrates this for `ReadFn` (a splittable `DoFn` that reads
+Avro files) processing the element `foo.avro` with restriction `[30, 70)`. This
+`@ProcessElement` call scans the Avro file for [data
+blocks](https://avro.apache.org/docs/current/spec.html#Object+Container+Files)
+starting from offset `30` and claims the position of each block in this range.
+If a block is claimed successfully, then the call outputs all records in this
+data block, otherwise, it terminates.
+
+<img class="center-block"
+    src="{{ site.baseurl }}/images/blog/splittable-do-fn/blocks.png"
+    alt="Processing a restriction by claiming blocks inside it"
+    width="400">
+
+For more details, see the [Restrictions, blocks and
+positions](http://s.apache.org/splittable-do-fn#heading=h.vjs7pzbb7kw).
+
+### Code example
+
+Let us look at some examples of SDF code. The examples use the Beam Java SDK,
+which [represents splittable
+`DoFn`s](https://github.com/apache/beam/blob/f7e8f886c91ea9d0b51e00331eeb4484e2f6e000/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java#L527)
+as part of the flexible [annotation-based
+`DoFn`](http://s.apache.org/a-new-dofn) machinery:
+
+* A splittable `DoFn` is a `DoFn` - no new base class needed. Any SDF derives
+from the `DoFn` class and has a `@ProcessElement` method.
+
+* The `@ProcessElement` method takes an additional
+[`RestrictionTracker`](https://github.com/apache/beam/blob/f7e8f886c91ea9d0b51e00331eeb4484e2f6e000/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/RestrictionTracker.java)
+parameter that gives access to the current restriction in addition to the
+current element.
+
+* An SDF needs to define a `@GetInitialRestriction` method that can create a
+restriction describing the complete work for a given element.
+
+* There are several other less important optional methods.
+
+The "Hello World" of SDF is a counter, which takes pairs *(x, N)* as input and
+produces pairs *(x, 0), (x, 1), …, (x, N-1)* as output.
+
+```java
```

Comment:
Is the Python version still to early to even give a tentative example of what it will look like? (Also, the restriction to Java should be called out below similar to the restriction to a subset of runners.)